src/ReductionMsg.chpl

```diff
@@ -32,253 +33,347 @@ module ReductionMsg
+                }
+                when "is_locally_sorted" {
+                  coforall loc in Locales with (&& reduce s) do on loc {
+                    const localSliceDom = sliceDom[eIn.a.localSubdomain()];
+                    for axisIdx in axes do
+                      s &&= isSortedOver(eIn.a, localSliceDom, axisIdx);
+                  }
                 }
+                otherwise halt("unreachable");
+              }
+              eOut.a[sliceIdx] = s;
             }
-            otherwise {
-                var errorMsg = unrecognizedTypeError(pn, dtype2str(gEnt.dtype));
-                rmLogger.error(getModuleName(),getRoutineName(),getLineNumber(),errorMsg);       
-                return new MsgTuple(errorMsg, MsgType.ERROR);       
+
+            const repMsg = "created " + st.attrib(rname);
+            rmLogger.info(getModuleName(),pn,getLineNumber(),repMsg);
+            return new MsgTuple(repMsg, MsgType.NORMAL);
+          }
+        }
+      }
+
+      select gEnt.dtype {
+        when DType.Int64 do return computeReduction(int);
+        when DType.UInt64 do return computeReduction(uint);
+        when DType.Float64 do return computeReduction(real);
+        when DType.Bool do return computeReduction(bool);
+        otherwise {
+          var errorMsg = notImplementedError(pn,dtype2str(gEnt.dtype));
+          rmLogger.error(getModuleName(),pn,getLineNumber(),errorMsg);
+          return new MsgTuple(errorMsg,MsgType.ERROR);
+        }
+      }
+    }
+
+    /*
+      Compute an array reduction along one or more axes.
+      (where the result has an integer data type)
+
+      Supports: 'argmin', 'argmax'
+    */
+    @arkouda.registerND(cmd_prefix="reduce->idx")
+    proc idxReductionMsg(cmd: string, msgArgs: borrowed MessageArgs, st: borrowed SymTab, param nd: int): MsgTuple throws {
+      use SliceReductionOps;
+      param pn = Reflection.getRoutineName();
+      const x = msgArgs.getValueOf("x"),
+            op = msgArgs.getValueOf("op"),
+            axis = msgArgs.get("axis").getPositiveIntValue(nd),
+            rname = st.nextName();
+
+      var gEnt: borrowed GenSymEntry = getGenericTypedArrayEntry(x, st);
+
+      if !idxReductionOps.contains(op) {
+        const errorMsg = notImplementedError(pn,op,gEnt.dtype);
+        rmLogger.error(getModuleName(),pn,getLineNumber(),errorMsg);
+        return new MsgTuple(errorMsg, MsgType.ERROR);
+      }
+
+      proc computeReduction(type t): MsgTuple throws {
+        const eIn = toSymEntry(gEnt, t, nd);
+
+        if nd == 1 {
+          var s: int;
+          select op {
+            when "argmin" {
+              const (minVal, minLoc) = minloc reduce zip(eIn.a, eIn.a.domain);
+              s = minLoc;
             }
+            when "argmax" {
+              const (maxVal, maxLoc) = maxloc reduce zip(eIn.a, eIn.a.domain);
+              s = maxLoc;
+            }
+            otherwise halt("unreachable");
```

Comment:
We usuaully return a msgtuple with the error type to avoid shutting down the whole server on this type of thing (we never want the server to crash or halt because that means all the data is lost).

Reply:
There is a check up above (line 257) that will return an error if the operation isn't recognized so the halt shouldn't be reachable (I had to put it there to make the compiler happy).

I think it would be clearer to return the error message from the select statement instead of checking beforehand, but returning from within a `forall` isn't allowed, so that's ruled out for the second select statement in this function. Do you think it would be better to refactor the code for the first select statement at least and leave the second as-is, or is it okay?

Reply:
OK, that makes sense to me, sousnds good as-is.